lightning/src/ln/channelmanager.rs

```diff
@@ -2297,9 +2308,11 @@ where
 		}
 	}
 
-	fn construct_recv_pending_htlc_info(&self, hop_data: msgs::OnionHopData, shared_secret: [u8; 32],
-		payment_hash: PaymentHash, amt_msat: u64, cltv_expiry: u32, phantom_shared_secret: Option<[u8; 32]>) -> Result<PendingHTLCInfo, ReceiveError>
-	{
+	fn construct_recv_pending_htlc_info(
+		&self, hop_data: msgs::OnionHopData, shared_secret: [u8; 32], payment_hash: PaymentHash,
+		amt_msat: u64, cltv_expiry: u32, phantom_shared_secret: Option<[u8; 32]>,
+		counterparty_skimmed_fee_msat: Option<u64>, chan_retrieval_info: ChanMapRetrieval
```

Comment:
> ISTM we can just pass allow_underpay rather than looking up the channel - in the update_add_htlc case we can move the ECH + onion_utils::decode_next_payment_hop call out of decode_update_add_htlc_onion and then take the per_peer_state mutex and look up the channel we're processing the message for.

Seems like that would move a lot of checks/codepaths to happen under the `Channel` lock, and it seems nicer to continue holding the lock for the minimal amount of time like we're currently doing? Or do you mean we'd still do all the forwarded-htlc checks outside of the channel lock, and only do the receiving checks inside of it? Btw currently we'll only look up the channel if an HTLC is actually underpaying.

Reply:
I had to keep the forwarding checks where they are because we can't hold two inner `PeerState` locks at the same time. Check it out plz.